ossl/src/lib.rs

```diff
@@ -0,0 +1,2378 @@
+    fn from(_error: std::num::TryFromIntError) -> Error {
+        Error::new(ErrorKind::WrapperError)
+    }
+}
+
+impl From<std::io::Error> for Error {
+    /// Maps an io error to a generic error
+    fn from(_error: std::io::Error) -> Error {
+        Error::new(ErrorKind::WrapperError)
+    }
+}
+
+#[cfg(all(feature = "log", feature = "fips"))]
+pub fn ossl_err_stack() -> String {
+    /* there is no external error management with fips builds */
+    "".to_string()
+}
+
+#[cfg(all(feature = "log", not(feature = "fips")))]
+pub fn ossl_err_stack() -> String {
+    // Use a mem bio to "print out" the error stack
+    let mut bio = std::ptr::null_mut();
+    let bio_method = unsafe { BIO_s_mem() };
+    if !bio_method.is_null() {
+        bio = unsafe { BIO_new(bio_method) };
+    }
+    if bio.is_null() {
+        return "Failed to fetch OpenSSL Error Stack".to_string();
+    }
+    unsafe { ERR_print_errors(bio) };
+
+    /* retrieve the mem bio data as a long string, with embedded \n */
+    let mut raw_mem: *mut c_char = std::ptr::null_mut();
+    let raw_len = unsafe {
+        BIO_ctrl(bio, BIO_CTRL_INFO as c_int, 0, void_ptr!(&mut raw_mem))
+    };
+
+    if raw_mem.is_null() || raw_len == 0 {
+        return "Failed to get error from OpenSSL's Error Stack".to_string();
+    }
+
+    /* copy this buffer to a vector so we can turn it into a Rust String */
+    let mut vec = vec![0u8; raw_len as usize];
+    unsafe {
+        std::ptr::copy_nonoverlapping(
+            raw_mem as *const u8,
+            vec.as_mut_ptr(),
+            vec.len(),
+        );
+    }
+
+    // remove final newline if any
+    while vec[vec.len() - 1] == b'\n' {
+        let _ = vec.pop();
+    }
+    match String::from_utf8(vec) {
+        Ok(s) => s,
+        Err(e) => format!("Failed to parse OpenSSL Error Stack: [{:?}]", e),
+    }
+}
+
+macro_rules! trace_ossl {
+    ($name:expr) => {
+        #[cfg(feature = "log")]
+        {
+            use log::error;
+            error!(
+                "{}:{}: {} failed: [{}]",
+                file!(),
+                line!(),
+                $name,
+                ossl_err_stack()
+            );
+        }
+    };
+}
+
+/// A structure representing the main crypto library context
+pub struct OsslContext {
+    context: *mut OSSL_LIB_CTX,
+}
+
+impl OsslContext {
+    pub fn new_lib_ctx() -> OsslContext {
+        OsslContext {
+            context: unsafe { OSSL_LIB_CTX_new() },
+        }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn from_ctx(ctx: *mut OSSL_LIB_CTX) -> OsslContext {
+        OsslContext { context: ctx }
+    }
+
+    pub fn ptr(&self) -> *mut OSSL_LIB_CTX {
+        self.context
+    }
+}
+
+unsafe impl Send for OsslContext {}
+unsafe impl Sync for OsslContext {}
+
+/// Wrapper around OpenSSL's `BIGNUM` type for handling large numbers.
+/// Manages the lifecycle and provides conversion methods.
+#[derive(Debug)]
+struct BigNum {
+    bn: *const BIGNUM,
+}
```

Comment:
Given that we are moving stuff around, i would probably move the separate wrappers to separate files as this file goes over 2k lines.